build.rs

```diff
@@ -0,0 +1,62 @@
+use anyhow::{bail, Context};
+use std::io::ErrorKind;
+use std::process::{Command, Stdio};
+
+fn main() -> anyhow::Result<()> {
+    println!("cargo::rerun-if-changed=.git/HEAD");
+    println!("cargo::rerun-if-changed=.git/refs");
+    let pkg_version = getenv("CARGO_PKG_VERSION")?;
+    match get_commit_hash()? {
+        Some(commit) => {
+            println!("cargo::rustc-env=GIT_COMMIT={commit}");
+            println!("cargo::rustc-env=VERSION_WITH_GIT={pkg_version} (commit: {commit})");
+        }
+        None => println!("cargo::rustc-env=VERSION_WITH_GIT={pkg_version}"),
+    }
+    Ok(())
+}
+
+fn get_commit_hash() -> anyhow::Result<Option<String>> {
+    let manifest_dir = getenv("CARGO_MANIFEST_DIR")?;
+    match Command::new("git")
+        .arg("rev-parse")
+        .arg("--git-dir")
+        .stdout(Stdio::null())
+        .stderr(Stdio::null())
+        .current_dir(&manifest_dir)
+        .status()
+    {
+        Ok(rc) if rc.success() => {
+            // We are in a Git repository
+            let output = Command::new("git")
+                .arg("rev-parse")
+                .arg("--short")
+                .arg("HEAD")
+                .current_dir(&manifest_dir)
+                .output()
+                .context("failed to run `git rev-parse --short HEAD`")?;
+            if !output.status.success() {
+                bail!(
+                    "`git rev-parse --short HEAD` command was not successful: {}",
+                    output.status
+                );
+            }
+            let revision = std::str::from_utf8(&output.stdout)
+                .context("`git rev-parse --short HEAD` output was not UTF-8")?
+                .trim()
+                .to_owned();
+            Ok(Some(revision))
+        }
+        Ok(_) => Ok(None), // We are not in a Git repository
```

Comment:
how does rust tells difference here from `Err(e)` below on when running not under git repo? 

it seems to just exit with non-0 

```
❯ git rev-parse --git-dir
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
❯ echo $?
128
```
so wouldn't it be an `Err`?

Reply:
`Command::status()` returns `Err` if the command couldn't be run at all (e.g., due to the program not being found) and returns `Ok` if it did run, even if it exited with a failure.  On line 29, we check whether the exit status was a success and fall through to line 50 if it wasn't.